services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/modules/nodeports.py

```diff
@@ -170,6 +170,76 @@ async def _get_data_from_port(port: Port) -> Tuple[Port, ItemConcreteValue]:
     return (port, ret)
 
 
+async def _download_files(
+    target_path: Path, download_tasks: Deque[Coroutine[Any, int, Any]]
+) -> Tuple[dict[str, Any], ByteSize]:
+    transferred_bytes = 0
+    data: dict[str, Any] = {}
+
+    if not download_tasks:
+        return data, ByteSize(transferred_bytes)
+
+    # TODO: limit concurrency to avoid saturating storage+db??
+    results: List[Tuple[Port, ItemConcreteValue]] = cast(
+        List[Tuple[Port, ItemConcreteValue]], await logged_gather(*download_tasks)
+    )
+    logger.info("completed download %s", results)
+    for port, value in results:
+
+        data[port.key] = {"key": port.key, "value": value}
+
+        if _FILE_TYPE_PREFIX in port.property_type:
+
+            # if there are files, move them to the final destination
+            downloaded_file: Optional[Path] = cast(Optional[Path], value)
+            dest_path: Path = target_path / port.key
+
+            if not downloaded_file or not downloaded_file.exists():
+                # the link may be empty
+                # remove files all files from disk when disconnecting port
+                logger.info("removing contents of dir %s", dest_path)
+                await remove_directory(
+                    dest_path, only_children=True, ignore_errors=True
+                )
+                continue
+
+            transferred_bytes = transferred_bytes + downloaded_file.stat().st_size
+
+            # in case of valid file, it is either uncompressed and/or moved to the final directory
+            logger.info("creating directory %s", dest_path)
+            dest_path.mkdir(exist_ok=True, parents=True)
+            data[port.key] = {"key": port.key, "value": str(dest_path)}
+
+            dest_folder = PrunableFolder(dest_path)
+
+            if _is_zip_file(downloaded_file):
+                # unzip updated data to dest_path
+                logger.info("unzipping %s", downloaded_file)
+                unarchived: Set[Path] = await unarchive_dir(
+                    archive_to_extract=downloaded_file, destination_folder=dest_path
+                )
+
+                dest_folder.prune(exclude=unarchived)
+
+                logger.info("all unzipped in %s", dest_path)
+            else:
+                logger.info("moving %s", downloaded_file)
+                dest_path = dest_path / Path(downloaded_file).name
+                await async_on_threadpool(
+                    # pylint: disable=cell-var-from-loop
+                    lambda: shutil.move(str(downloaded_file), dest_path)
+                )
+
+                dest_folder.prune(exclude={dest_path})
```

Comment:
Q: what is the point of having a dest_folder and prune it by excluding itself? what is suppose to happen?

Reply:
Yes it's a bit of an odd one here. Let's say you previously had a file connected to the input named `a.txt`. Now you replace it with a file called `b.txt`. You would want to delete previously created `a.txt` file.
I will add a comment here.

Reply:
:+1: